internal/logging/sentry_replacer.go

```diff
@@ -45,13 +52,56 @@ func NewSentryReplacer(w io.Writer) *SentryReplacer {
 }
 
 func (sr *SentryReplacer) Write(p []byte) (n int, err error) {
+	if sr.closed {
+		return 0, io.EOF
+	}
+
 	sr.mu.Lock()
 	defer sr.mu.Unlock()
 
+	// Append p to our own buffer, see if there's anything we need to censor.
+	sr.buf = append(sr.buf, p...)
+
+	// If we've appended at least a line, censor it and write it out.
+	for {
+		if len(sr.buf) == 0 {
+			// Buffer flushed out completely.
+			return len(p), nil
+		}
+
+		idx := bytes.IndexRune(sr.buf, '\n')
+		if idx < 0 {
+			// No line yet, just lie to the caller and tell them we wrote p.
+			return len(p), nil
+		}
+
+		var line []byte
+		line, sr.buf = sr.buf[:idx+1], sr.buf[idx+1:]
+		line = sr.re.ReplaceAll(line, replacement)
+
+		_, err := sr.w.Write(line)
+		if err != nil {
+			// This is not strictly the error related to the incoming `p`, but the best we can do.
+			return 0, fmt.Errorf("cannot filter: %w", err)
+		}
+	}
+}
+
+// implements zerolog.LevelWriter
+func (sr *SentryReplacer) WriteLevel(level zerolog.Level, p []byte) (n int, err error) {
 	if sr.closed {
 		return 0, io.EOF
 	}
 
+	// skip the filter unless we will actually write to sentry
+	if _, ok := sr.levels[level]; !ok {
+		//nolint:wrapcheck
+		return sr.w.WriteLevel(level, p)
+	}
+
+	sr.mu.Lock()
+	defer sr.mu.Unlock()
+
```

Comment:
Is there a reason to do this before locking a mutex? I am not quite sure I am getting the goal of this PR.
But it seems like we are working with code achievable from goroutines. We have a pointer here, and executing several sr.w.WriteLevel on the same pointer in different threads might eventually make a mess.
Or is it not the case?

Reply:
The reason is that if the requested level is not in enabled levels, we just pass the call down to the underlying logger.

The main reason for the mutex here is the internal buffer where we keep the logs in between calls. Tho to be fair I'm not sure why its needed and I'm not  sure how the logger works, but IMO when we do not touch that buffer, we do not need to worry about locking.

The underlying logger should IMO be thread safe by its own. It's only a theory tho, so maybe @lzap can chim in here :)

Reply:
Yeah the mutex is only needed for the buffer which is shared.

From the zerolog docs, it seems that the library does not affect thread safety, it guarantees that every single log event = single write which is thread safe as long as your underlaying writer is thread safe (e.g. appending to a file in Linux).